src/parallel_ops/reduction.cc

```diff
@@ -37,6 +38,23 @@ using Legion::Task;
 using Legion::TaskArgument;
 using Legion::TaskLauncher;
 
+/* Params */
+bool operator==(ReductionParams const &lhs, ReductionParams const &rhs) {
+  return lhs.reduction_legion_dim == rhs.reduction_legion_dim &&
+         lhs.reduction_degree == rhs.reduction_degree;
+}
+
+bool ReductionParams::is_valid(ParallelTensorShape const &input) const {
+  return input.is_valid();
```

Comment:
Missing some constraints here

Reply:
Same comment as above for validity checks.